src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs

```diff
@@ -141,7 +141,11 @@ public class Frame<TContext> : Frame
                         }
                     }
 
-                    Reset();
+                    // Don't lose request rejection state
+                    if (!_requestRejected)
+                    {
+                        Reset();
```

Comment:
Now I see why my above comment was wrong. I'll keep it there in case there isn't a test for it yet. If there was a test that failed that caused you to add this if, just mention it.